File: fern/snippets/tool-router/python/in-chat-auth.py
import asyncio
from composio import Composio
from agents import Agent, Runner, HostedMCPTool

composio = Composio(api_key="your-composio-api-key")

async def chat():
    print("Creating Tool Router session...")
    session = await composio.create("pg-user-550e8400-e29b-41d4")
    print(f"Tool Router session created: {session.mcp.url}")

    agent = Agent(
        name="GitHub Assistant",
        instructions="Help users with their GitHub repositories.",
        tools=[
            HostedMCPTool(
                tool_config={
                    "type": "mcp",
                    "server_label": "composio",
                    "server_url": session.mcp.url,
                    "require_approval": "never",
                    "headers": {
                        "x-api-key": "your-composio-api-key",
                    },
                }
            )
        ],
    )

    print("Chat with your agent. Type 'exit' to quit.\n")

    while True:
        user_input = input("You: ")
        if user_input.lower() == "exit":
            break

        result = await Runner.run(
            agent,
            user_input,
        )

        print(f"Agent: {result.final_output}\n")

asyncio.run(chat())

File: fern/snippets/tool-router/typescript/in-chat-auth.ts
import { openai } from "@ai-sdk/openai";
import { experimental_createMCPClient as createMCPClient } from "@ai-sdk/mcp";
import { Composio } from "@composio/core";
import { generateText } from "ai";
import * as readline from "readline";

const composio = new Composio({ apiKey: "your-composio-api-key" });

async function chat() {
  console.log("Creating Tool Router session...");
  const { mcp } = await composio.create("pg-user-550e8400-e29b-41d4");
  console.log(`Tool Router session created: ${mcp.url}`);

  const client = await createMCPClient({
    transport: {
      type: "http",
      url: mcp.url,
      headers: {
        "x-api-key": "your-composio-api-key",
      },
    },
  });

  const tools = await client.tools();
  const messages: { role: "user" | "assistant"; content: string }[] = [];

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  console.log("Chat with your agent. Type 'exit' to quit.\n");

  const prompt = () => {
    rl.question("You: ", async (input) => {
      if (input.toLowerCase() === "exit") {
        rl.close();
        return;
      }

      messages.push({ role: "user", content: input });

      const { text } = await generateText({
        model: openai("gpt-4o"),
        tools,
        messages,
        maxSteps: 5,
      });

      console.log(`Agent: ${text}\n`);
      messages.push({ role: "assistant", content: text });

      prompt();
    });
  };

  prompt();
}

chat();
